### ISSUE TITLE
Consider promoting Wyam console output to the web server/browser in debug/dev/watch mode

### ISSUE THREAD

**Issue by avishnyakov** [statiqdev/Statiq#157]
it seems that Wyam processing errors can be divided into two groups:
- Fatal errors, so that the web site cannot be started and the whole flow is down
- Minor, non-fatal errors, so that the web site is started but some files are locked or something

Wyam.exe has a fantastic output showing if all good or there are some errors.
The point is that we have 3 entry points while athoring something w/ Wyam in --watch mode:
1) The IDE, we edit files
2) Wyam.exe, keep an eye on processing issues/errors
3) The browser, a few tabs in the browser so we can see what's up with the final rendering

That flow is a bit  inefficient.
First of all, we need to reload pages in the browser once Wyam is done processing. There is a separate ticket on that - Consider automatic page reload once Wyam regenerated new pages #155 

Next, if there is a minor error like some files are locked and can't be copies/written (js/css/fonts), then you might spend some time trying to figure out why the browser displays some spoiled content. You go back to the Wyam console and oh, yep, here we go. That means every time you edit file in the IDE you have to go to Wyam console, check there is no errors, and then go to the browser. 

Things get worse if there is a major error and Wyam can't process site. You edit the file, go to the browser -> refresh the page and get "This web page is not available". You like, c'mon.. go back to the Wyam console again, ah... here we go.. typo or missed assembly or whatever.

So the point is to improve that flow, the 1-2-3 steps eliminating the need to go back to console and see what's there.

Maybe Wyam would be able to inject some kind of status bar, a notification or something into the final web site to indicate minor errors? The web site is up and running, but there are minor error in the pipeline so Wyam would push an status bar at the top of the web site, some kind of notification status saying "hey, I did the web site but there are some issues - wanna see details? here we are"

And maybe Wyam would be able not only crash in the console with the fatal error / exception, but also start the web site redirecting all pages to the page with the latest  latest exception? Like asp.net does with its "yellow" exception page, and so Wyam would start the web site and show the exception trace.

Generally speaking, replicating the Wyam console output in the web browser would be interesting move with lots of potential opportunities and goodies. Like asp.net trace page, the general concept. 

That also would help with CI/CD integrity. Say we have CI deploying the web site once the code checked-in - how do we know if all was cool and solid, no issues were there while Wyam processing? Pushed to the github, it goes to Appveyor, and all gets published. But was the processing done well? Any outcome in the prod web site so we can be sure all was good?

All that just some ideas raised up while intensively working with Wyam, authoring master pages / layout pages with Razor and some content. The authoring flow can be seen in (1) ide (2) Wyam console an (3) the web browser and yet to be improved. Eliminating the need to switch between IDE/console/browser and moving all things from the console to the browser in some way would be a valuable improvement on the authoring flow efficiency.

**Comment by daveaglick** [comment]
Love, love, love where you're going with this. Something like Glimpse hooked into the engine. It would have to be generic and opt-in since Wyam may not be generating web sites at all. Let me think about this one for a little while to consider what would work best...

**Comment by daveaglick** [comment]
Been giving this some thought. Here's what I'm thinking right now:
- This should work by injecting a reference to a JavaScript file into all pages delivered via the embedded OWIN web server.
- The real work (including modifying the DOM to display a GUI) should be done by the JavaScript to keep the modification to the page HTML to a minimum (same way MiniProfiler, Glimpse, etc. work).
- Should consider splitting the preview web server capability off into a separate library so it's not tied to the console application (though I don't want it in Wyam.Core either given the large number of OWIN dependencies).
- This new server library should require a reference to the engine (which shouldn't know anything about the server). It will register a trace handler as well as get any other information needed from the engine.
- Communication with the client should be via SignalR if possible.
- Injection should be on by default, but capable of turning off (add a `--no-injection` flag to console application).
- Should include automatic page reload on rebuild (also on by default but able to turn off, add a `--no-reload` flag to console application).
- All trace output from most recent build should be visible from the widget (with option to turn verbose output on or off).
- A button should be available within the widget to manually trigger a rebuild (and automatically reload the page when complete if that option is turned on).
- The widget should proactively report warnings and errors (call-out, toast message, etc.).

**Comment by avishnyakov** [comment]
Sounds good. Hope you wouldn't overengineer it :)
